cmd/pprofessor: introduce pprof-based tool

axw opened the pull request:
## Motivation/summary

pprofessor is based on the pprof tool, using an alternative
"Fetcher" implementation that queries Elasticsearch to
aggregate continuous profiling samples recorded by APM Server.

I haven't gone through all of the pprof flags to make sure
they work, so some may not make sense. There are a couple of
flags added specifically for our use case:

* `-service`, which defines the service name to filter on.
* `-start`, which defines the start timestamp (or date math)
  from which to start aggregating. If specified along with
  `-seconds`, the latter  controls the end time. If only `-start`
  is specified, then the end time is now. If only `-seconds` is
  specified, then the end time is now and the start time is duration
  seconds before now.

We record the number of profiles and documents aggregated in
profile comments. All metrics (currently: cpu, inuse/allocated
heap objects/space) are aggregated into a single profile.

To aggregate duration correctly, we introduce the `profile.id`
field, which is a unique ID generated per profile, shared by
all sample docs derived from a profile.

No automated tests for now, and comes with no warranty or support.

Note that there is a known issue with how the Go Agent reports "alloc"
samples: it reports ever-increasing counters, rather than deltas. We
need to change this, so we can visualise the allocations within a define
time range. See https://github.com/elastic/apm-agent-go/issues/708

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
- [ ] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

- Run APM Server with CPU and Heap profiling enabled in APM Server:

`apm-server -E apm-server.instrumentation.enabled=true -E apm-server.instrumentation.profiling.cpu.enabled=true -E apm-server.instrumentation.profiling.cpu.interval=10s -E apm-server.instrumentation.profiling.heap.enabled=true`

- Run pprofessor:

`go run ./cmd/pprofessor -start=now-1h -service=apm-server --http=:6060 http://admin:changeme@localhost:9200`

## Demo

![pprofessor3](https://user-images.githubusercontent.com/843579/88249642-a23a8300-ccd7-11ea-9d74-f4c75aa6a79c.gif)

## Related issues

Closes #3828

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-4017/detail/PR-4017/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-4017/detail/PR-4017/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-4017/detail/PR-4017/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-4017/detail/PR-4017/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #4017 opened]`

* Reason: `Aborted from #2`

* Start Time: 2020-07-23T03:30:54.689+0000

* Duration: 3 min 19 sec


* Commit: a682e303cd18430d062f0efc0d9da0885924541b




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Check out from version control`
    * Description: [2020-07-23T03:33:39.543Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-07-23T03:33:
    
    * Duration: 0 min 8 sec
    
    * Start Time: 2020-07-23T03:33:36.861+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-4017/runs/1/steps/11/log)
    
  
  * **Name**: `Check out from version control`
    * Description: [2020-07-23T03:34:05.154Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-07-23T03:34:
    
    * Duration: 0 min 5 sec
    
    * Start Time: 2020-07-23T03:34:05.130+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-4017/runs/1/steps/15/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-07-23T03:33:45.037Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-07-23T03:33:45.037Z] Caused: hudson.plugins.git.GitException: Could not checkout a682e303cd18430d062f0efc0d9da0885924541b
[2020-07-23T03:33:45.037Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2767)
[2020-07-23T03:33:45.037Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
[2020-07-23T03:33:45.037Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:154)
[2020-07-23T03:33:45.037Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
[2020-07-23T03:33:45.037Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2020-07-23T03:33:45.037Z] 	at hudson.remoting.Request$2.run(Request.java:369)
[2020-07-23T03:33:45.037Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
[2020-07-23T03:33:45.037Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-07-23T03:33:45.037Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-07-23T03:33:45.037Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-07-23T03:33:45.037Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:93)
[2020-07-23T03:33:45.037Z] 	at java.lang.Thread.run(Thread.java:748)
[2020-07-23T03:33:45.037Z] Retrying
[2020-07-23T03:33:45.108Z] Sleeping for 20 sec
[2020-07-23T03:33:44.097Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2020-07-23T03:33:44.097Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-23T03:33:44.118Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-server.git +refs/pull/4017/head:refs/remotes/origin/PR-4017 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-07-23T03:33:44.994Z]  > git config core.sparsecheckout # timeout=10
[2020-07-23T03:33:45.009Z]  > git checkout -f a682e303cd18430d062f0efc0d9da0885924541b # timeout=15
[2020-07-23T03:34:05.154Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-07-23T03:34:05.233Z] Wiping out workspace first.
[2020-07-23T03:34:05.266Z] Cloning the remote Git repository
[2020-07-23T03:34:05.266Z] Using shallow clone with depth 3
[2020-07-23T03:34:05.266Z] Avoid fetching tags
[2020-07-23T03:34:05.279Z] Cloning repository git@github.com:elastic/apm-server.git
[2020-07-23T03:34:05.279Z]  > git init /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-4017 # timeout=10
[2020-07-23T03:34:05.298Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2020-07-23T03:34:05.298Z]  > git --version # timeout=10
[2020-07-23T03:34:05.314Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-23T03:34:05.321Z]  > git fetch --no-tags --progress -- git@github.com:elastic/apm-server.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2020-07-23T03:34:09.044Z] Cleaning workspace
[2020-07-23T03:34:09.069Z] Using shallow fetch with depth 3
[2020-07-23T03:34:09.069Z] Pruning obsolete local branches
[2020-07-23T03:34:09.023Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2020-07-23T03:34:09.030Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-07-23T03:34:09.039Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2020-07-23T03:34:09.051Z]  > git rev-parse --verify HEAD # timeout=10
[2020-07-23T03:34:09.064Z] No valid HEAD. Skipping the resetting
[2020-07-23T03:34:09.064Z]  > git clean -fdx # timeout=10
[2020-07-23T03:34:09.076Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2020-07-23T03:34:09.076Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-23T03:34:09.081Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-server.git +refs/pull/4017/head:refs/remotes/origin/PR-4017 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-07-23T03:34:09.906Z] Merging remotes/origin/master commit 886d11f7b4327653e3a034ca383e8d41d899fe64 into PR head commit a682e303cd18430d062f0efc0d9da0885924541b
[2020-07-23T03:34:09.951Z] ERROR: Execution failed
[2020-07-23T03:34:09.951Z] hudson.plugins.git.GitException: Command "git checkout -f a682e303cd18430d062f0efc0d9da0885924541b" returned status code 128:
[2020-07-23T03:34:09.951Z] stdout: 
[2020-07-23T03:34:09.951Z] stderr: fatal: reference is not a tree: a682e303cd18430d062f0efc0d9da0885924541b
[2020-07-23T03:34:09.951Z] 
[2020-07-23T03:34:09.951Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2430)
[2020-07-23T03:34:09.951Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$1100(CliGitAPIImpl.java:81)
[2020-07-23T03:34:09.951Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2743)
[2020-07-23T03:34:09.951Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from apm-ci-immutable-ubuntu-1804-1595475071473967763.c.elastic-ci-prod.internal/10.224.1.21:33302
[2020-07-23T03:34:09.951Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1788)
[2020-07-23T03:34:09.951Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
[2020-07-23T03:34:09.951Z] 		at hudson.remoting.Channel.call(Channel.java:998)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:146)
[2020-07-23T03:34:09.951Z] 		at sun.reflect.GeneratedMethodAccessor1123.invoke(Unknown Source)
[2020-07-23T03:34:09.951Z] 		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2020-07-23T03:34:09.951Z] 		at java.lang.reflect.Method.invoke(Method.java:498)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:132)
[2020-07-23T03:34:09.951Z] 		at com.sun.proxy.$Proxy114.execute(Unknown Source)
[2020-07-23T03:34:09.951Z] 		at jenkins.plugins.git.MergeWithGitSCMExtension.checkout(MergeWithGitSCMExtension.java:144)
[2020-07-23T03:34:09.951Z] 		at jenkins.plugins.git.MergeWithGitSCMExtension.decorateRevisionToBuild(MergeWithGitSCMExtension.java:110)
[2020-07-23T03:34:09.951Z] 		at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1063)
[2020-07-23T03:34:09.951Z] 		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1168)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
[2020-07-23T03:34:09.951Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2020-07-23T03:34:09.951Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2020-07-23T03:34:09.951Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-07-23T03:34:09.952Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-07-23T03:34:09.952Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-07-23T03:34:09.952Z] Caused: hudson.plugins.git.GitException: Could not checkout a682e303cd18430d062f0efc0d9da0885924541b
[2020-07-23T03:34:09.952Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2767)
[2020-07-23T03:34:09.952Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
[2020-07-23T03:34:09.952Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:154)
[2020-07-23T03:34:09.952Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
[2020-07-23T03:34:09.952Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2020-07-23T03:34:09.952Z] 	at hudson.remoting.Request$2.run(Request.java:369)
[2020-07-23T03:34:09.952Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
[2020-07-23T03:34:09.952Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-07-23T03:34:09.952Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-07-23T03:34:09.952Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-07-23T03:34:09.952Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:93)
[2020-07-23T03:34:09.952Z] 	at java.lang.Thread.run(Thread.java:748)
[2020-07-23T03:34:09.952Z] Retrying
[2020-07-23T03:34:10.028Z] Sleeping for 30 sec
[2020-07-23T03:34:09.913Z]  > git config core.sparsecheckout # timeout=10
[2020-07-23T03:34:09.918Z]  > git checkout -f a682e303cd18430d062f0efc0d9da0885924541b # timeout=15
[2020-07-23T03:34:12.601Z] Running on worker-854309 in /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-4017
[2020-07-23T03:34:12.744Z] [39;49m[INFO] getVaultSecret: Getting secrets[0m
[2020-07-23T03:34:12.863Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-07-23T03:34:15.177Z] + chmod 755 generate-build-data.sh
[2020-07-23T03:34:15.177Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-4017/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-4017/runs/1 ABORTED 199061
[2020-07-23T03:34:15.177Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-4017/runs/1/steps/?limit=10000 -o steps-info.json
[2020-07-23T03:34:18.357Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-4017/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-07-23T03:34:19.065Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-4017/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

jalvz commented:
This doesn't seem to work for me with the 7.8 stack:

```http://admin:changeme@localhost:9200: [400 Bad Request] {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"}],"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"},"status":400}```

A few other notes/questions:

- Non obvious requirements for running this should be documented (min stack version, Graphviz, etc) 
- The command semantics are not so intuitive to me: 
  Since this is mostly for us, I'd expect `service` to be hardcoded to `apm-server`, or at least default to it. 
  The behaviour of specifying only `seconds` is conventional?  it doesn't look so useful..
  I think the Elasticsearch URL should have its own flag, as other required arguments have it too.  
- While not tested, I'd remove all non-essential code, like  tls_ca handling.   
- The expression "unique profile ID shared by all samples" is not so intuitive, consider `profileKey`, `profileGroupID`, or similar.
- It would be more coherent to use types for both Elasticsearch responses and queries or none, but not only responses.
- I think Elasticsearch types should live on their own package, or at least their own file.

axw commented:
@jalvz thanks for the review!

> http://admin:changeme@localhost:9200: [400 Bad Request] {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"}],"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"},"status":400}

That's probably because there's a composite aggregation on `profile.id`, which is introduced in this PR. You would need to build and run apm-server to use the tool. (Part of the reason for creating this tool is to surface missed data/requirements such as this.)

> Non obvious requirements for running this should be documented (min stack version, Graphviz, etc)

I can add a basic README. Until then: my intention was for the tool to be used with the same (branch) stack & server version. Otherwise, no additional dependencies apart from pprof's (which includes Graphviz).

> The command semantics are not so intuitive to me:
> Since this is mostly for us, I'd expect service to be hardcoded to apm-server, or at least default to it.

I'll set "apm-server" as the default.

> The behaviour of specifying only seconds is conventional? it doesn't look so useful..

That's one of the standard `pprof` flags. If you run `go tool pprof -seconds=60 http://service:1234/debug/pprof/profile` it'll profile for 60 seconds.

> I think the Elasticsearch URL should have its own flag, as other required arguments have it too.

This is part of the pprof tool: the "source" (i.e. Elasticsearch URL in this case) is a positional argument.

> While not tested, I'd remove all non-essential code, like tls_ca handling.

Sure, I'll remove it until it's needed.

> The expression "unique profile ID shared by all samples" is not so intuitive, consider profileKey, profileGroupID, or similar.

Are you referring to the wording in `fields.yml`? I'll try to clarify the description if that's what you mean. I don't think we should change the field name. It is the ID of the profile to which the samples belong. Parent/child (1:n) relationship.

> It would be more coherent to use types for both Elasticsearch responses and queries or none, but not only responses.

The reason for having struct types for responses is because we need to operate on the results. The requests are just JSON marshalled, so it doesn't really matter if they're structs or maps. I did what's simplest.

> I think Elasticsearch types should live on their own package, or at least their own file.

-1 on separate package, see https://dave.cheney.net/practical-go/presentations/qcon-china.html#_consider_fewer_larger_packages

If you think it's really worthwhile I can create a separate file, but I don't think it is. The types are used in exactly one place.

axw commented on cmd/pprofessor/fetcher.go:
```diff
@@ -0,0 +1,442 @@
+}
+
+// Fetch fetches a profile from Elasticsearch, by aggregating samples for a specified service name and time range.
+func (f *fetcher) Fetch(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
```
Done

jalvz commented:
> You would need to build and run apm-server to use the tool

Don't follow. Did `make` and `make update`, then ran with relevant flags. Anything else needed?

> no additional dependencies apart from pprof's (which includes Graphviz).

Hmm, I actually had to apt-install it...

axw commented:
>> You would need to build and run apm-server to use the tool
>
> Don't follow. Did make and make update, then ran with relevant flags. Anything else needed?

Looking at the error message again, maybe I introduced a bug in between testing and proposing this PR. I'll look into it and come back.

>> no additional dependencies apart from pprof's (which includes Graphviz).
>
> Hmm, I actually had to apt-install it...

Sorry, what I meant was "no additional dependencies other than pprof's dependencies, among which is Graphviz" -- not that it's included within the pprof binary.

axw commented:
> Looking at the error message again, maybe I introduced a bug in between testing and proposing this PR. I'll look into it and come back.

I hadn't tested under the scenario where there was an existing index/template without the `profile.id` field. If you remove your index and overwrite the template then it should work. Either way, I've added a bug fix which will avoid the error noted.

I've added a README and (hopefully) clarified the meaning of `profile.id`.

axw commented:
Merging without a changelog entry, as this is not intended for end users.
